Add architecture.md document, api.md and pull request template

Ullaakut opened the pull request:
## Goal of this PR

Fixes #4 

This PR adds an `architecture.md` document that describes internal components, their interactions, and details the database _schema_ in the form of tables describing the formatting of keys and values in the Badger database.

It also adds an `api.md` file which is referenced by the `architecture.md` document. This document details the REST API only for now, as it is the only one that is already available.

Since both of those documents REQUIRE maintenance as we introduce changes to the codebase, either in the design of the system or in the details of our APIs, this PR also adds a pull request template which aims at ensuring that opened PRs update the documentation accordingly. It contains a very light check list for now, as we don't yet exactly know what we will want this checklist to include.

## Checklist

- [x] Is on the right branch
- [x] Documentation is up-to-date
- [x] ~Tests are up-to-date~

## Additional Notes

For now this PR is a draft, because it is missing some elements. We need to add a license in order for the project to be indexable by pkg.go.dev and for us to be able to reference it in our documentation. If we end up using a license which does not allow indexing (extremely unlikely), we will be able to locally run godoc and to generate the documentation ourselves instead, but for now this PR assumes that we will be indexed by pkg.go.dev.

Just like #6 , this PR adds a `docs` folder at the root of the repository. Once both are merged, or if this PR gets merged before #6, it would probably be nice to include links to `architecture.md` and `api.md` from the `introduction.md` document as well.

awfm9 commented on docs/api.md:
````diff
@@ -0,0 +1,132 @@
+```json
+{
+  "height": 425,
+  "key": "20VDBpq8GZS3JDdmskD70Deq+kdsNetzeCZA=",
+  "value": "+Em4QOD6BU2j9jPdks07DsmalPEDfbsdAk/Cdvs0Dkawo/ei02ocds9D9klSflsa8CH8p9A4ID6A6A"
````
This in not correct, they are both hex-encoded. I made that change a while back because I missed it, but it was always supposed to be. Sorry if you missed it.

awfm9 commented on docs/api.md:
```diff
@@ -0,0 +1,132 @@
+
+### `GET /values/:encoded_key` - Get Value
+
+This route returns the payload value of an encoded Ledger entry.
```
This route returns the Ledger payload for a Ledger key from the execution state trie. The encoding of the key is inspired by the canonical string format for keys, but made compatible with URLs and with requests for multiple URLs.

awfm9 commented on docs/api.md:
```diff
@@ -0,0 +1,132 @@
+
+#### Query Parameters
+
+* `hash`: Optional. Specifies which commit hash to get the payload value from. Defaults to the latest commit from the state.
```
You can use `state commitment hash` when you want to be a bit more precise in the docs. Defaults to the commit for the last sealed block that the service indexed.

awfm9 commented on docs/api.md:
```diff
@@ -0,0 +1,132 @@
+#### Query Parameters
+
+* `hash`: Optional. Specifies which commit hash to get the payload value from. Defaults to the latest commit from the state.
+* `version`: Optional. Specifies the pathfinding version to use to traverse the state. Defaults to the default pathfinder key encoding.
```
Defaults to the default pathfinder *path* encoding used in the respective version of the Flow Go dependency.

awfm9 commented on docs/api.md:
````diff
@@ -0,0 +1,132 @@
+
+```json
+[
+  "6c49490a1f023fda632cfe3a49b66201",
+  "24e32f4633ff12daf66f1e2d8c73b04f",
+  "7bc1e622a5b639e8befe97262d3a21c5",
+  "1d7dd90eca1066a5905abf243b926d35",
+  "8c5178bcaa7b30cec5c9073aee1a1702"
````
These can be of variable size, so maybe it would be good to illustrate that.

awfm9 commented on docs/architecture.md:
```diff
@@ -0,0 +1,125 @@
+
+### Database Schema
+
+The DPS uses [BadgerDB](https://github.com/dgraph-io/badger) to store datasets of state changes and block information.
```
...to build all of the indexes required for random protocol and execution state access. It doesn't re-use any of the protocol state database, but instead re-indexes everything, so that all databases used to bootstrap can be discarded subsequently.

awfm9 commented on docs/architecture.md:
```diff
@@ -0,0 +1,125 @@
+|:-------------------|:------------------|:-----------|
+| **Type**           | byte              | hex hash   |
+| **Description**    | Index type prefix | Block ID   |
+| **Example Value**  | `1`               | `1FD5532A` |
```
I don't like upper case hex strings :D

Ullaakut replied:
They probably don't like you either! :p Will change it!

awfm9 reviewed (commented):
I added some comments, but the architecture and storage part need updating.

Ullaakut commented:
Will update to reflect the changes that came since #23

Ullaakut commented:
Rebased and added:

* New indexes
* New Rosetta components
* Removed some outdated stuff

It's now ready for a new quick review @awfm9 and a merge if it sounds good to you :p

awfm9 reviewed (approved):
I fixed some small problems myself, and updates the ToC on each document to be consistent (it's automatically generated by my extension).
